Added Tab Support

milaiwi opened the pull request:
Changes:

- Added tabs when opening a file
- Added an EmptyPage when no file is currently selected
- Cleaned up codebase
- Moved settingsModal, newNoteModal, newDirectoryModal, and flashcardModeModal to a Provider. This makes it simplier to implement (EmptyPage, NewNote, and IconsSidebar all need these fields) without props passing hell. Also simplifies adding keyboard shortcuts in the future (just import useModalOpener and set fields to true).
- Fixed overflow issues
- Moved modals to a Provider
- Added animations on opening tab

Removed:

- Had to remove newNote and newDirectory fields on the contextMenu for file sidebar that got displayed when right clicking. Reason for this is there was no *clean* way of making this work without calls between backend and frontend which defeated the purpose of moving all of tabs and modals to a Provider. I'll add it in the future to the frontend which makes a final call to backend for the filesystem. Another added benefit of moving it away from the backend is for custom UI.
- ReplaceWord will be a separate PR


Note:
- Reason for changing openRelativePath instead of openAbsolutePath in newNote and newDirectory:

1.  We always have access to filePath which is an absolute path which makes it much more conveniant. 
2. Creating a newNote or newDirectory stores it on the directory whose tab is currently opened (i.e., if you are currently viewing /notes/test1/t1.md and create a newNote called t2, then it will be written to /notes/test1/t2.md)

- Reason for `window.electronStore.removeOpenTabsByPath(path)` inside `use-file-by-filepath` which then sends an event to the provider:
1) If you delete a file, the tab should get automatically closed. The only other way to make this work without calls to backend would be to wrap `MainPageComponent` inside `TabProvider`. However, this goes against `useContext` since it would re-render the entire app on any change to openTabs (which is a lot). The occasional cost to the backend is worth it!

Questions
- Does suggest word still not work after making it a boolean field instead of string? It works for me.
- Let me know if you can find any more issues (overflow, paths, etc..)!

samlhuillier commented on src/components/Sidebars/TabSidebar.tsx:
```diff
@@ -0,0 +1,190 @@
+/* Displays the filepath when hovering on a tab */
+const Tooltip: React.FC<TooltipProps> = ({ filepath, position }) => {
+  return createPortal(
+    <div className="tab-tooltip" style={{ top: `${position.y}px`, left: `${position.x}px` }}>
```
I think for a follow on PR but noting this now: we need to handle things like resizing tabs on window resize, a scrollbar for tabs and things like minimum width for tabs

samlhuillier reviewed (changes requested):
Overall looks great @milaiwi! Thanks for updating the PR. I have added a couple of comments where I think it could be improved slightly

milaiwi commented:
I'll take a look and respond to your requests!

milaiwi commented on src/components/File/hooks/use-file-by-filepath.ts:
```diff
@@ -107,6 +107,31 @@ const useFileByFilepath = () => {
     openFileByPath(absolutePath)
   }
 
+  const openAbsolutePath = async (filePath: string, optionalContentToWriteOnCreate?: string): Promise<void> => {
```
> I don't think we need this function. It seems to be very similar to the openRelative except the check for whether its an absolute path.

The issue I run into is that openRelative takes the path and joins it with the vault directory. The issue is that the `filePath` state we store is the `absolutePath` which means if we want to easily open a file, we'd have to first remove the vaultDirectory section then call it. This is tedious for every call. Instead this function just simplifies the logic since our ipcHandler stores absolutePaths and we pass through our prop an absolutePath. I think this is a fine function.

> I think we should just have a generic function which opens path (both absolute and relative) and then we should perform the check for whether its an absolute path with an ipc handler and the node path.isAbsolute() function

Sure this could work. However, `path.isAbsolute` does not work properly on windows since files begin with a drive instead of `/`

joseplayero replied:
Ok fair enough. Have you tested this on windows?

samlhuillier replied:
Ok fair enough. Have you tested on Windows to confirm that it doesn't work? Because from this it seems like node:path behaviour depends on the host machine: https://nodejs.org/api/path.html#windows-vs-posix

milaiwi commented on src/utils/strings.ts:
```diff
@@ -51,3 +51,13 @@ export const getInvalidCharacterInFileName = async (filename: string): Promise<s
+ * Returs true if the filePath is absolute. path.isAbsolute does not correctly work
+ * with Windows.
+ */
+export const isFilePathAbsolute = (filePath: string): boolean => {
```
As mentioned above, `path.isAbsolute` returns false for `C:\Users\user\Documents\hello.txt`

samlhuillier commented on src/styles/global.css:
```diff
@@ -45,7 +53,7 @@ button {
 
 ::-webkit-scrollbar {
-  width: 7px;
+  width: 2px;
```
thoughts @milaiwi ?

samlhuillier commented on src/components/File/FileSideBar/index.tsx:
```diff
@@ -84,17 +84,12 @@ const FileExplorer: React.FC<FileExplorerProps> = ({
 
   return (
-    <div onContextMenu={handleMenuContext} className="h-full grow">
+    <div className="h-full grow p-6 pr-0 opacity-70">
```
the left padding should also be 0 here

milaiwi commented:
@samlhuillier Let me know if you find any more issues. 

- Fixed Padding
- Fixed issue with tab width
- Cleaned up settings to work on different screen sizes (probably not an appropriate PR to do this, but it was raised as an issue in this PR so I went ahead and fixed it)

samlhuillier commented:
Thanks @milaiwi two final things:
- Could you increase the width of the scrollbar in the file sidebar it's difficult to click right now
- Could you respond to the isAbsolute issue:
> Have you tested on Windows to confirm that it doesn't work? Because from this it seems like node:path behaviour depends on the host machine: https://nodejs.org/api/path.html#windows-vs-posix

milaiwi commented:
> Could you increase the width of the scrollbar all scrollbars are too small to click

Should be 7px now. Let me know if that's good

> Could you respond to the isAbsolute issue (if you don't have a windows laptop I can test on windows tomorrow):

It should work now. I changed openAbsolutePath and deleted the function that originally did it. I tested on Windows and it works. I'm pretty sure everything should be good now, but let me know
